### DISCUSSION TITLE
Overload resolution only picks the first

### DISCUSSION THREAD

**Discussion by InSyncWithFoo** [microsoft/pyright#8096] [General]
Consider the following example ([playground](https://pyright-play.net/?pyrightVersion=1.1.366&pythonVersion=3.13&strict=true&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwFgAoF0SKJAT1y1IHNseAkQCCpLgBooABXBIwAYzAAbaWABuaECrDEAJiyPNFK4vXpQAyglz5C9ALIIVAbQDy0gCoBdABRylEqqAJQAXCxQUVD6aMCwMBAucH70aCrA6kgAFtphUJ5QAPQhUAC0AHxQXvkAdPXGpuaWNnYi9ABKSW6FvgHywSrhkdGx8XAg3SlpGVm5IPmFJeVVNVD1tcYsTRZQAJKkSBHM0VAAApraugYjUWMJUzCp6ZlQlPP5B0jSy5X7h3UGidohctDo9IZgXc4g9kk8Zq93nkoAAxa7fYqlP5ovRHdZA24xGFwR7PWZvHLI8RSTErKDUwGbVjMnaWHHUY6ne4kuFkxGUhb-IgAH1R6J%2BWKq7LxGy2zFBVwhLHu3VcACFvP5iPlWvYkE4XOrNdIAEb5NWS6qMlgK8E3Zgqw2iTV%2BbX00062x6zqq53VHyWtay5mOlSu-LUj30iSWhn4pnGdBaYgqGDcXBoPxfPylABUQpzpVOxYAxEKWEm0Cm0zxM9m82LcYXi9Ey9KK2hk6n05npTmoPn60WW230R2uzWM34%2Bw2Z8PW43qInO1Xu7W-N0s4cc9Ih-OomWvuPV5PM5v69I5-Oy9pwIKAETS%2B-YSx8ZR4ahYE0qNBQADuWA5FAuADMoKhQPeurtF0hpfN4ZzdD497HtWPYbi407ojuBYhLhpyjriKFrlOm4zpeWG4UWBFLswQA)):

```python
class SupportsMul[O, T](Protocol):
    def __mul__(self, other: O, /) -> T: ...

class SupportsRmul[O, T](Protocol):
    def __rmul__(self, other: O, /) -> T: ...
```

```python
class Int:
    @overload
    def __mul__(self, other: Int, /) -> Int: ...
    @overload
    def __mul__(self, other: Float, /) -> Float: ...

    def __mul__(self, other: Any, /) -> Any: ...

class Float:
    def __mul__(self, other: Int | Float, /) -> Float: ...
```

```python
@overload
def mul[B, T](a: SupportsMul[B, T], b: B) -> T: ...
@overload
def mul[A, T](a: A, b: SupportsRmul[A, T]) -> T: ...

def mul(a: Any, b: Any) -> Any: ...
```

```python
reveal_type(Int() * Int())          # Int
reveal_type(Int() * Float())        # Float
reveal_type(Float() * Int())        # Float
reveal_type(Float() * Float())      # Float

reveal_type(mul(Int(), Int()))      # Int
reveal_type(mul(Int(), Float()))    # error: "Float" is incompatible with protocol "SupportsRmul[Int, T@mul]"
reveal_type(mul(Float(), Int()))    # Float
reveal_type(mul(Float(), Float()))  # Float
```

If the overloads were to be ordered differently, the results would change accordingly ([playground](https://pyright-play.net/?pyrightVersion=1.1.366&pythonVersion=3.13&strict=true&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwFgAoF0SKJAT1y1IHNseAkQCCpLgBooABXBIwAYzAAbaWABuaECrDEAJiyPNFK4vXpQAyglz5C9ALIIVAbQDy0gCoBdABRylEqqAJQAXCxQUVD6aMCwMBAucH70aCrA6kgAFtphUJ5QAPQhUAC0AHxQXvkAdPXGpuaWNnYi9ABKSW6FvgHywSrhkdGx8XAg3SlpGVm5IPmFJeVVNVD1tcYsTRZQAJKkSBHM0VAAApraugYjUWMJUzCp6ZlQlPP5AGLXSNLLlVBvnojusGidohctDo9IZwXc4g9kk8Zq93nl9oc-qUAQcQRtjKd7nBHs9Zm8cujxFJitiqlS6mDGmZdkDqMdCQjiUjSaiKQsMUQAD6An5YlYi4EMzasGWQq4wlj3bquABC3n8xHyrXsSCcLlV6ukACN8iradUpSw5dCbswlfrROq-JqoI6oCbrLYdZ1lW7fOa1viZfaVM78lTjeGJOb6aDpcZ0FpiCoYNxcGg-Li-KUAFQC7OlU5FgDEApYibQydTPAzWdzEuoBaL0VLrKQ5bQSZTaYzbezUDzdcLzdbPw7Xer6b8ffrM%2BHLYb7ZlFarPb83Uzh2z0iH86ipe04H5ACJccfsJY%2BMo8NQsEaVGgoAB3LA5KC4AbKFRQY-a9pdfVcW8M5uh8Y9x0rbsa3XFxNyQbdFwLedR2BCDV2gjcZx3LcQlw04UOoNCoKnTCfgQudCwIpcgA)):

```python
class Int:
    @overload
    def __mul__(self, other: Float, /) -> Float: ...
    @overload
    def __mul__(self, other: Int, /) -> Int: ...

    def __mul__(self, other: Any, /) -> Any: ...
```

```python
reveal_type(mul(Int(), Int()))      # error: "Int" is incompatible with protocol "SupportsRmul[Int, T@mul]"
reveal_type(mul(Int(), Float()))    # Float
reveal_type(mul(Float(), Int()))    # Float
reveal_type(mul(Float(), Float()))  # Float
```

I expected Pyright to check all overloads. Why wasn't this the case?

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug. Mypy produces the same results. Refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=overloads) for a description of pyright's overload matching algorithm.

The problem with your "error" case is that the provided arguments do not work with either overload signature. You can more easily debug this by creating separate (non-overload) signatures.

```python
def mul1(a: SupportsMul[B, T], b: B) -> T: ...
def mul2(a: A, b: SupportsRmul[A, T]) -> T: ...

mul1(Int(), Float()) # Error
mul2(Int(), Float()) # Error
```

**Reply by InSyncWithFoo** [reply]
I think `(a: SupportsMul[B, T], b: B) -> T:` should match, with `Float` as both `B` and `T`. The second overload is clearly the only "winner", since `Int` is not assignable to `Float`.

Shouldn't [SFIAE](https://en.wikipedia.org/wiki/Substitution_failure_is_not_an_error) apply in this case?

**Reply by erictraut** [reply]
Ah, I see what you mean. I agree that theoretically this should work, but in practice it won't work with pyright's (or mypy's) current constraint solver implementation. The problem is that when evaluating the expression `Int()` with the inference context `SupportsMul[B, T]`, there is insufficient context to guide the algorithm to select the second overload and produce `SupportsMul[Float, Float]`, so the expression `Int()` is evaluated as `SupportsMul[Int, Int]`. 

The typing spec provides no guidance for type checker authors when it comes to bidirectional type inference, constraint solving, or overload resolution. This code involves all three of these areas.

Interestingly, if you swap the two arguments, pyright's constraint solver is able to find the solution, but mypy's (still) cannot.

```python
def mul1(a: SupportsMul[B, T], b: B) -> T: ...
mul1(b=Float(), a=Int())
```

This is because pyright's algorithm performs multiple passes and attempts to refine the inference context based on the partially-solved type variables during each pass.

I'm considering rewriting pyright's constraint solver to be more like mypy's. While pyright's current implementation can handle some edge cases that mypy cannot (like the example above where the arguments are swapped), pyright's approach is very complicated and difficult to maintain, is computationally much more expensive, and produces results that are dependent on argument ordering.

Regardless of whether I stick with the current algorithm or adopt mypy's, the code in your original sample above will not type check without errors.
